### PR TITLE
Adding 1st version of the Design model

### PR DESCRIPTION
**Pull request by 2potatocakes**
Apologies for the large PR - there's a lot of specs and a few fixture files in here though ;)

This PR includes all the code required for indexing `Designs` in its currently proposed state, I included the feedback on the current PR that @dferdian has up to discuss the shape of the schema here: https://github.com/envato/websites-catalogue/pull/10

**PLEASE NOTE:** The design schema is still under review though and still very likely to change in the coming days.

There should be no surprises in this code but there are a couple of things worth pointing out which differ slightly from our Hosted-Themes repo.

- Added support for `design` data to be stored as either an array of different designs within a single file or else just as a single hash of a single design. There are a couple of specs and some test files to  ensure that both formats are supported
- Also, I'd like to get feedback and see how people feel about this, but I wasn't expecting the `vendor` attribute to be set within the json supplied by hosted/sites. For the moment, I'm setting the `vendor` attribute for each design based upon the `data/designs/[vendor]` directory that the design lives within. How do people feel about this? I kind of always assumed that we'd eventually have some sort of authentication layer that would be able to set this value, but I never really assumed this value would be set within the `Design` data itself. Anyways, would love to get some feedback.

### CODE REVIEW

File: lib/catalogue/designs/data_source.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+module Catalogue
+  module Designs
+    class DataSource
+      def self.read
+        Dir.glob(WebsitesCatalogue.config.designs_data.join('**/*.json')).map do |path_to_design|
+          vendor = File.dirname(path_to_design).to_s.gsub(/.*\//,'')
+          design_json = JSON.parse(File.read(path_to_design), symbolize_names: true)
+          design_json = [design_json] unless design_json.is_a?(Array)
```
Comment (2potatocakes) [inline lib/catalogue/designs/data_source.rb]: ps. I don't love this bit.. if anyone has some better suggestions I'm all 👂's ?

**Comment by travo** [comment]
Damn, should have read this pull request before [this one](https://github.com/envato/websites-catalogue/pull/14).

File: config/environment.rb [hunk 2]
```diff
@@ -14,20 +14,24 @@ module WebsitesCatalogue
   ELASTICSEARCH_INDEX_NAME = 'websites_catalogue_v1'
 
   class Config
-    attr_writer :index_name, :categories_data, :designs_data
-
-    def index_name
-      @index_name ||= ELASTICSEARCH_INDEX_NAME
-    end
+    attr_writer :index_name, :categories_data, :designs_data, :designs_data
```
Comment (travo) [inline config/environment.rb]: You've got `designs_data` here twice buddy.

File: config/environment.rb [hunk 3]
```diff
@@ -14,20 +14,24 @@ module WebsitesCatalogue
...
 
     def elasticsearch_url
       @elasticsearch_url ||= secret(:elasticsearch_url) || ENV['ELASTICSEARCH_URL']
     end
 
+    def index_name
+      @index_name ||= ELASTICSEARCH_INDEX_NAME
+    end
+
```
Comment (travo) [inline config/environment.rb]: Did you mean to re-order the methods? It just adds a little extra diff-noise.

File: lib/catalogue/designs/data_source.rb (hunk 1 above)
Comment (travo) [inline lib/catalogue/designs/data_source.rb]: Yeah, @dferdian has a [nice pattern](https://github.com/envato/hosted-themes/pull/91/files#diff-14a1442d730dd98538ab15c18e7e430b) for taking a class method like this and using it to instantiate an instance to access private methods. You can then refactor some of this stuff out to private methods and clean up the `read` method.

File: lib/catalogue/designs/data_source.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
+module Catalogue
+  module Designs
+    class DataSource
+      def self.read
+        new.read
+      end
+
+      def read
+        designs_paths.map { |path| processed_json(path) }.flatten
```
Comment (blimpage) [inline lib/catalogue/designs/data_source.rb]: `flat_map` will do the same job as `map.flatten`, and [can often be faster](https://gist.github.com/PragTob/8f037b6cab4bffca411063aced6a7917).

File: lib/catalogue/designs/data_source.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+      private
+
+      def processed_json(path)
+        design_data(path).map { |json| json.merge(vendor: vendor_name_from_path(path)) }
+      end
+
+      def design_data(path)
+        design_json = JSON.parse(File.read(path), symbolize_names: true)
+        design_json.is_a?(Array) ? design_json : [design_json]
```
Comment (dferdian) [inline lib/catalogue/designs/data_source.rb]: have you tried `[design_json].flatten` ? I think it's better than having object type checking

File: lib/ingestion/schema.rb [hunk 6]
```diff
@@ -21,6 +23,62 @@ def category_mappings
...
+          }
+        },
+        'author' => {
+          'type' => 'object',
+          'properties' => {
+            'name' => Types::TEXT,
+            'url'  => Types::TEXT,
+          }
+        },
+        'plans'             => {
```
Comment (dferdian) [inline lib/ingestion/schema.rb]: unnecessary spaces?

File: lib/catalogue/designs/data_source.rb (hunk 5 above)
Comment (2potatocakes) [inline lib/catalogue/designs/data_source.rb]: 🤔  Nice suggestion, cheers

File: spec/catalogue/categories/serializer_spec.rb [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  }
+
+  describe ".serialize" do
+    subject(:serialized_category) { described_class.serialize(category) }
+
+    it "serializes a correct name" do
+      expect(serialized_category[:name]).to eq("Magic Card Tricks")
+    end
+
+    context "case sensitive UUID's" do
```
Comment (blimpage) [inline spec/catalogue/categories/serializer_spec.rb]: Nitpick: this apostrophe is unnecessary. So is the one in the `it` description below

Comment (2potatocakes) [inline spec/catalogue/categories/serializer_spec.rb]: I always thought acronyms were supposed to use an apostrophe for plurals, turns out I was wrong, cheers @blimpage ❤️

File: spec/catalogue/designs/data_source_spec.rb [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+          expect(data_source.detect {|design| design[:vendor] == 'sneaky-alternate-vendor-name'}).to be_nil
+        end
+      end
+    end
+
+    describe "a design" do
+      let(:avada_creative_uuid) { '445e4bf1-987b-4b72-a5f5-ca4bf31a0b7f' }
+      subject(:design) { data_source.detect {|design| design[:uuid] == avada_creative_uuid} }
+
+      it "has the expected uuid" do
```
Comment (blimpage) [inline spec/catalogue/designs/data_source_spec.rb]: This whole `describe "a design"` spec could just be a single `it` block that was:
```
expect(design).to eq({
  uuid: "foo",
  vendor: "bar",
  ...etc
})
```
That'd be shorter _and_ it'd catch cases where there are additional unexpected attributes in the hash.

File: spec/catalogue/designs/data_source_spec.rb [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    subject(:data_source) { described_class.read }
+
+    describe "assigns a vendor name" do
+      describe "multiple designs" do
+        it "reads all 5 designs in from both Hosted and Sites" do
+          expect(data_source.size).to eq(5)
+        end
+
+        it "reads 2 designs from the Hosted team" do
+          expect(data_source.select {|design| design[:vendor] == 'hosted'}.size).to eq(2)
```
Comment (dferdian) [inline spec/catalogue/designs/data_source_spec.rb]: another nitpick, you might need to replace `select + size` with `count`
```
expect(data_source.count {|design| design[:vendor] == 'hosted'}).to eq(2)
```

File: lib/catalogue/designs/data_source.rb [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+      def design_data(path)
+        [JSON.parse(File.read(path), symbolize_names: true)].flatten
+      end
+
+      def designs_paths
+        Dir.glob(WebsitesCatalogue.config.designs_data.join('**/*.json'))
+      end
+
+      def vendor_name_from_path(path)
+        path.match(/^#{WebsitesCatalogue.config.designs_data}\/(sites|hosted)\//i)[1]
```
Comment (blimpage) [inline lib/catalogue/designs/data_source.rb]: Now we've got our list of allowed vendors hardcoded in this regex 😞 

Maybe we could do something less fancy but more adaptable? e.g.
 - Remove `"#{WebsitesCatalogue.config.designs_data}\/"` from the path
 - Split path by `"\/"`
 - Grab `[0]` from the resulting array

If we do then want to validate that against a known list of vendors, we could store that in a constant and then `raise unless LIST_OF_KNOWN_VENDORS.include?(vendor)`

Comment (2potatocakes) [inline lib/catalogue/designs/data_source.rb]: Yeah sorry @blimpage, my bad, wasn't thinking straight before, just realised we didn't support nested files and was trying to make sure we supported nested .json files underneath `data/designs/[vendor]/` and added that in 🙏

Comment (2potatocakes) [inline lib/catalogue/designs/data_source.rb]: my first attempt was using this:
```
path.match(/^#{WebsitesCatalogue.config.designs_data}\/([^\/]*)\//i)[1]
```
Which does the same thing as stripping the designs path and grabbing the first folder after that. I like the exception idea

Comment (2potatocakes) [inline lib/catalogue/designs/data_source.rb]: Are you ok with this @blimpage? https://github.com/envato/websites-catalogue/pull/12/commits/38985ebc010c52f95abd349275e6e38701f6fb59

Comment (blimpage) [inline lib/catalogue/designs/data_source.rb]: Looks rad to me! :tada:

Comment (2potatocakes) [inline lib/catalogue/designs/data_source.rb]: Included a spec for it to: https://github.com/envato/websites-catalogue/pull/12/commits/cc1db69c23ce3ed9420953b067c6594d8f5bd112

**Pull request merged**